Platform/OpenNtfsDxe/Info.c

```diff
@@ -0,0 +1,245 @@
+    return Status;
+  }
+
+  if (!BaseMftRecord->InodeRead) {
+    BaseMftRecord->FileRecord = AllocateZeroPool (FileSystem->FileRecordSize);
+    if (BaseMftRecord->FileRecord == NULL) {
+      DEBUG ((DEBUG_INFO, "NTFS: Failed to allocate buffer for FileRecord.\n"));
+      FreeFile (BaseMftRecord);
+      FreePool (BaseMftRecord);
+      return EFI_OUT_OF_RESOURCES;
+    }
+
+    Status = ReadMftRecord (BaseMftRecord->File, BaseMftRecord->FileRecord, BaseMftRecord->Inode);
+    if (EFI_ERROR (Status)) {
+      DEBUG ((DEBUG_INFO, "NTFS: Could not read Mft Record #%d.\n", BaseMftRecord->Inode));
+      FreeFile (BaseMftRecord);
+      FreePool (BaseMftRecord);
+      return Status;
+    }
+  }
+
+  Status = InitAttr (&BaseMftRecord->Attr, BaseMftRecord);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_INFO, "NTFS: Could not initialize attribute.\n"));
+    FreeFile (BaseMftRecord);
+    FreePool (BaseMftRecord);
+    return Status;
+  }
+
+  Res = (ATTR_HEADER_RES *)FindAttr (&BaseMftRecord->Attr, AT_VOLUME_NAME);
+  if ((Res != NULL) && (Res->NonResFlag == 0) && (Res->InfoLength != 0)) {
+    //
+    // The Volume Name is not terminated with a Unicode NULL.
+    // Its name's length is the size of the attribute as stored in the header.
+    //
+    if (Res->InfoLength > (MAX_FILE_SIZE - sizeof (CHAR16))) {
+      DEBUG ((DEBUG_INFO, "NTFS: Volume Name is too huge.\n"));
+      FreeFile (BaseMftRecord);
+      FreePool (BaseMftRecord);
+      return EFI_VOLUME_CORRUPTED;
+    }
+
+    *Label = AllocateZeroPool (Res->InfoLength + sizeof (CHAR16));
+    if (*Label == NULL) {
+      DEBUG ((DEBUG_INFO, "NTFS: Failed to allocate buffer for *Label\n"));
+      FreeFile (BaseMftRecord);
+      FreePool (BaseMftRecord);
+      return EFI_OUT_OF_RESOURCES;
+    }
+
+    CopyMem (*Label, (UINT8 *)Res + Res->InfoOffset, Res->InfoLength);
+  }
+
+  FreeFile (BaseMftRecord);
+  FreePool (BaseMftRecord);
+
+  return EFI_SUCCESS;
+}
+
+/**
+  Returns file info, system info, or the label of a volume
+  depending on information type specified.
+
+  @param  This   Pointer to this instance of file protocol
+  @param  Type   Pointer to information type requested
+  @param  Len    Pointer to size of buffer
+  @param  Data   Pointer to buffer for returned info
+
+  @retval EFI_STATUS  Status of the operation
+**/
+EFI_STATUS
+EFIAPI
+FileGetInfo (
+  IN EFI_FILE_PROTOCOL  *This,
+  IN EFI_GUID           *Type,
+  IN OUT UINTN          *Len,
+  OUT VOID              *Data
+  )
+{
+  EFI_STATUS                    Status;
+  EFI_NTFS_FILE                 *File;
+  EFI_FILE_INFO                 *Info;
+  EFI_FILE_SYSTEM_INFO          *FSInfo;
+  EFI_FILE_SYSTEM_VOLUME_LABEL  *VLInfo;
+  CHAR16                        *Label;
+  UINTN                         Length = 0;
+
+  ASSERT (This != NULL);
+  ASSERT (Type != NULL);
+  ASSERT (Len != NULL);
+  ASSERT ((Data != NULL) || ((Data == NULL) && (*Len == 0)));
```

Comment:
No need to state mutually exclusive conditions? i.e. whether `Data` can be `NULL`.

If it can be NULL, then `Data` should be `OPTIONAL` as well.

Reply:
As for `Data` declaration, it should follow specification.
```
typedef
EFI_STATUS
(EFIAPI *EFI_FILE_GET_INFO) (
IN EFI_FILE_PROTOCOL * This ,
IN EFI_GUID *InformationType,
IN OUT UINTN *BufferSize ,
OUT VOID *Buffer
);

Reply:
The definitions are vague as usual. The issue here is `Buffer` may be `NULL`, but the call can never succeed when `Buffer` is `NULL`. Imo follow the spec and leave as-is.